### PR TITLE
trigger service auth /login endpoint

### PR DESCRIPTION
**Pull request by aherrmann-da**
Implements the `/login` endpoint of the trigger service authentication middleware as specified [here](https://github.com/digital-asset/daml/blob/7c038119aade97fcfea1753efc5aca4e5837ed68/triggers/service/authentication.md).
Adds a test-case for the `/login` endpoint following the example of the oauth-test-service test-cases.
Part of https://github.com/digital-asset/daml/issues/6923.

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import scala.language.postfixOps
+import scala.util.Try
+import spray.json._
+
+// This is an implementation of the trigger service authentication middleware
+// for OAuth2 as specified in `/triggers/service/authentication.md`
+object Server extends StrictLogging {
+  import com.daml.oauth.server.JsonProtocol._
+
+  private def toRedirectUri(uri: Uri) =
```
Comment (cocreature) [inline triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala:30]: I think we might want to make the authority here or probably even better just the full URL configurable. Consider a case where this is running behind NGINX. We might be listening on `localhost` but the external facing host will be different and the path could also be different (e.g. it could be behind some prefix).

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    val requests = TrieMap[UUID, Uri]()
+    val route = concat(
+      path("auth") {
+        get {
+          complete((StatusCodes.NotImplemented, "The /auth endpoint is not implemented yet"))
+        }
+      },
+      path("login") {
+        get {
+          parameters(('redirect_uri.as[Uri], 'claims))
```
Comment (cocreature) [inline triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala:49]: Maybe worth extracting the request bodies into separate functions? Totally fine to leave that for separate PRs.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+              login =>
+                extractRequest {
+                  request =>
+                    val requestId = UUID.randomUUID
+                    requests += (requestId -> login.redirectUri)
+                    val authorize = OAuthRequest.Authorize(
+                      responseType = "code",
+                      clientId = config.clientId,
+                      redirectUri = toRedirectUri(request.uri),
+                      scope = Some(login.claims),
```
Comment (cocreature) [inline triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala:60]: We probably need to make the scopes configurable to some degree. Some oauth providers (e.g. auth0) also allow you to use custom parameters instead of scopes to accomplish the same thing. Not quite sure what the best UX for that is and we don’t have to solve this in this PR but something to keep in mind for the future.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                extractRequest {
+                  request =>
+                    val redirectUri = for {
+                      state <- authorize.state
+                      requestId <- Try(UUID.fromString(state)).toOption
+                      redirectUri <- requests.remove(requestId)
+                    } yield redirectUri
+                    redirectUri match {
+                      case None =>
+                        complete(StatusCodes.NotFound)
```
Comment (cocreature) [inline triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala:85]: Maybe unauthorized would be a more appropriate status code?

File: triggers/service/auth/src/test/scala/com/daml/oauth/middleware/Test.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+          assert(resp.status == StatusCodes.Found)
+          val req = HttpRequest(uri = resp.header[Location].get.uri)
+          Http().singleRequest(req)
+        }
+      } yield {
+        // Redirect to CALLBACK
+        assert(resp.status == StatusCodes.Found)
+        assert(resp.header[Location].get.uri == Uri("http://CALLBACK"))
+        // Store token in cookie
+        assert(resp.header[`Set-Cookie`].get.cookie.name == "token")
```
Comment (cocreature) [inline triggers/service/auth/src/test/scala/com/daml/oauth/middleware/Test.scala:44]: Should we also assert on the value of the cookie? Might also be a good idea to use something slightly more specific like “daml-ledger-token” (or something else).

**Review by cocreature — APPROVED** [review]
Great work, thank you very much! A few comments but nothing that necessarily needs to be addressed in this PR.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Config.scala [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    // OAuth2 client properties
+    clientId: String,
+    clientSecret: String,
+)
+
+private[middleware] object Config {
+  private val Empty =
+    Config(
+      port = Port.Dynamic,
+      oauthUri = Uri().withScheme("http"),
```
[the remark this answers is not in the document]
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Config.scala:23]: It allows to implement the `--oauth-host` and `--oauth-port` flags concisely. Those flags are also marked as required, so we won't end up with just `http://` from the command-line flags. Is there a better way to do that?

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Config.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        .action((x, c) => c.copy(oauthUri = c.oauthUri.withPort(x)))
+        .required()
+        .text("Port of the OAuth2 server")
+
+      opt[String]("id")
+        .action((x, c) => c.copy(clientId = x))
+        .required()
+        .text("OAuth2 client id")
+
+      opt[String]("secret")
```
[the remark this answers is not in the document]
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Config.scala:55]: Good point, that part is copied from the test client. I'll switch it to use envvars.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Request.scala [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.oauth.middleware
+
+import akka.http.scaladsl.model.Uri
+import spray.json.{DefaultJsonProtocol, JsString, JsValue, JsonFormat, deserializationError}
+
+object Request {
+
+  case class Login(redirectUri: Uri, claims: String) // TODO[AH] parse ledger claims
```
[the remark this answers is not in the document]
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Request.scala:11]: It's the URI that the middleware should redirect to after the login flow completed. So, it's the URI of the original request. I'll add a comment pointing this out.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala (hunk 1 above)
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala:30]: Good point

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala (hunk 2 above)
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala:49]: Yeah, that's probably better for readability, especially once the other endpoints are filled in.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala (hunk 3 above)
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala:60]: That's good to know, the implementation of that is probably best guided by a concrete use-case that requires that sort of thing. The mapping from claims to scope generally still needs work.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala (hunk 4 above)
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala:85]: The idea behind 404 was that the login request under the given `state` was not found. 401 seems problematic, the `/cb` is part of the authentication flow, not something that requires authentication itself, also 401 requires a challenge in the [WWW-Authenticate header field](https://tools.ietf.org/html/rfc7235#section-3.1) which doesn't seem appropriate here.

File: triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Config.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Config.scala:23]: Is there a reason why we accept `oauth-host` and `oauth-port` instead of accepting a single parameter that specifies a URI? Host and port makes sense for gRPC but for HTTP Requests it seems a bit weird.

[the remark this answers is not in the document]
Comment (aherrmann-da) [reply triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Config.scala:23]: Fair enough, no good reason, really. I'll make it a single flag.
